### PR TITLE
librustc_typeck: use diag item instead of string compare

### PR DESCRIPTION
**Pull request by da-x**

### CODE REVIEW

File: src/librustc_typeck/check/op.rs [hunk 1]
```diff
@@ -562,7 +562,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                    on the left and may require reallocation. This \
                    requires ownership of the string on the left";
 
-        let is_std_string = |ty| &format!("{:?}", ty) == "std::string::String";
+        let string_type = self.tcx.get_diagnostic_item(sym::string_type);
+        let is_std_string = |ty: Ty<'tcx>| match &ty.ty_adt_def() {
```
Comment (lcnr) [inline src/librustc_typeck/check/op.rs:566]:
```suggestion
        let is_std_string = |ty: Ty<'tcx>| match ty.ty_adt_def() {
```

I don't think this ref is needed, is it?

**Review by lcnr — APPROVED** [review]

File: src/librustc_typeck/check/op.rs [hunk 2]
```diff
@@ -562,7 +562,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                    on the left and may require reallocation. This \
                    requires ownership of the string on the left";
 
-        let is_std_string = |ty| &format!("{:?}", ty) == "std::string::String";
+        let string_type = self.tcx.get_diagnostic_item(sym::string_type);
+        let is_std_string = |ty: Ty<'tcx>| match ty.ty_adt_def() {
+            Some(ty_def) => Some(ty_def.did) == string_type,
+            None => false,
+        };
```
Comment (petrochenkov) [inline src/librustc_typeck/check/op.rs:569]: You could also use `map_or` here though.
```
ty.ty_adt_def().map_or(false, |adt_def| Some(adt_def.did) == string_type)
```

**Pull request merged**